### PR TITLE
Spawn restriction registry

### LINKED ISSUE #230: [Suggestion] SpawnRestrictionRegistry
SpawnRestriction is used to limit mob spawning to water or ground, also with the heightmap type.

Example syntax:
`SpawnRestrictionRegistry.INSTANCE.add(EntityType.PIG, SpawnRestriction.Location.ON_GROUND, HeightMap.Type.MOTION_BLOCKING_NO_LEAVES);`

### PR DESCRIPTION
**Pull request by YanisBft**
Very simple
Invoker by Linguardium on discord
Closes #230

### CODE REVIEW

**Comment by i509VCB** [comment]
Looks like we have a conflict lol, as my PR adds this to a subclass of the entity builder (#828)

File: fabric-object-builder-api-v1/src/main/java/net/fabricmc/fabric/api/object/builder/v1/entity/FabricSpawnRestrictionRegistry.java [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+	 *
+	 * @param type      the entity type
+	 * @param location  the environment type where the mob can spawn
+	 * @param heightmap the heightmap type
+	 * @param predicate a spawn predicate
+	 */
+	public static <T extends MobEntity> void register(EntityType<T> type, SpawnRestriction.Location location, Heightmap.Type heightmap, SpawnRestriction.SpawnPredicate<T> predicate) {
+		SpawnRestrictionAccessor.callRegister(type, location, heightmap, predicate);
+	}
+}
```
Comment (liach) [inline fabric-object-builder-api-v1/src/main/java/net/fabricmc/fabric/api/object/builder/v1/entity/FabricSpawnRestrictionRegistry.java:64]: private constructor

**Review by liach — COMMENTED** [review]
Also test mod preferred

**Comment by YanisBft** [comment]
> 
> 
> Looks like we have a conflict lol, as my PR adds this to a subclass of the entity builder (#828)

Ah my bad, didn't even look

**Comment by liach** [comment]
@i509VCB that's not really a conflict. also imo we should keep this pr since there is a point of registering spawn restriction aside from the entity type.

**Comment by i509VCB** [comment]
Anyways in the worst case I just inline my builder PR to call these

File: fabric-object-builder-api-v1/src/main/java/net/fabricmc/fabric/api/object/builder/v1/entity/FabricSpawnRestrictionRegistry.java (hunk 1 above)
Comment (YanisBft) [reply fabric-object-builder-api-v1/src/main/java/net/fabricmc/fabric/api/object/builder/v1/entity/FabricSpawnRestrictionRegistry.java:64]: doing a test mod would require adding an entity, which seems a bit too much

**Comment by liach** [comment]
I wonder if you can override existing values with this register hook.

**Review by liach — APPROVED** [review]

**Comment by YanisBft** [comment]
> I wonder if you can override existing values with this register hook.

You can't, the game crashes with "Duplicate registration for type minecraft:xxx"

File: fabric-object-builder-api-v1/src/main/java/net/fabricmc/fabric/api/object/builder/v1/entity/FabricSpawnRestrictionRegistry.java (hunk 1 above)
Comment (modmuss50) [reply fabric-object-builder-api-v1/src/main/java/net/fabricmc/fabric/api/object/builder/v1/entity/FabricSpawnRestrictionRegistry.java:64]: A test mod to add an entity might be useful to test other parts of the api.

**Review by modmuss50 — APPROVED** [review]
Looks good.

**Comment by modmuss50** [comment]
#828 might be a better solution however.

File: fabric-object-builder-api-v1/src/main/java/net/fabricmc/fabric/api/object/builder/v1/entity/FabricSpawnRestrictionRegistry.java [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+import net.fabricmc.fabric.mixin.object.builder.SpawnRestrictionAccessor;
+
+/**
+ * Allows registering spawn restrictions for mob entities.
+ */
+public final class FabricSpawnRestrictionRegistry {
+	/**
+	 * Private logger, not exposed.
+	 */
+	private static final Logger LOGGER = LogManager.getLogger();
```
Comment (valoeghese) [inline fabric-object-builder-api-v1/src/main/java/net/fabricmc/fabric/api/object/builder/v1/entity/FabricSpawnRestrictionRegistry.java:42]: Ah yes documenting loggers

File: fabric-object-builder-api-v1/src/main/java/net/fabricmc/fabric/api/object/builder/v1/entity/FabricSpawnRestrictionRegistry.java [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+import net.minecraft.entity.mob.MobEntity;
+import net.minecraft.util.registry.Registry;
+import net.minecraft.world.Heightmap;
+
+import net.fabricmc.fabric.mixin.object.builder.SpawnRestrictionAccessor;
+
+/**
+ * Allows registering spawn restrictions for mob entities.
+ */
+public final class FabricSpawnRestrictionRegistry {
```
Comment (valoeghese) [inline fabric-object-builder-api-v1/src/main/java/net/fabricmc/fabric/api/object/builder/v1/entity/FabricSpawnRestrictionRegistry.java:41]: Private constructor

File: fabric-object-builder-api-v1/src/main/java/net/fabricmc/fabric/api/object/builder/v1/entity/FabricSpawnRestrictionRegistry.java [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+	 *  EntityType.BAT,
+	 *  SpawnRestriction.Location.ON_GROUND,
+	 *  Type.MOTION_BLOCKING_NO_LEAVES,
+	 *  BatEntity::canSpawn);</pre>
+	 * </p>
+	 *
+	 * <p>Note: A mob can only have one spawn restriction registration.</p>
+	 *
+	 * @param type      the entity type
+	 * @param location  the environment type where the mob can spawn
```
Comment (valoeghese) [inline fabric-object-builder-api-v1/src/main/java/net/fabricmc/fabric/api/object/builder/v1/entity/FabricSpawnRestrictionRegistry.java]: “Environment type where the mob can spawn” sounds weird imo. Consider changing the documentation to just say “environment” for that

**Review by valoeghese — CHANGES_REQUESTED** [review]

**Review by Player3324 — APPROVED** [review]
Looks good, please move the example to the class javadoc and fix valoghese's comments before this can be merged

**Review by i509VCB — APPROVED** [review]

**Pull request closed without merging**